### PR TITLE
Add exchange UpBit #40

### PR DESCRIPTION
**Pull request by panevgen**
https://github.com/altangent/ccxws/issues/40
https://docs.upbit.com/docs/upbit-quotation-websocket

### CODE REVIEW

File: src/exchanges/upbit-client.js [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+    let { opening_price, trade_price, acc_bid_volume, acc_ask_volume, code, trade_volume, acc_trade_volume, change_rate, change_price, low_price, high_price, timestamp } = msg;
+    
+    let market = this._tickerSubs.get(code);
+    return new Ticker({
+      exchange: "Upbit",
+      base: market.base,
+      quote: market.quote,
+      timestamp,
+      last: trade_price,
```
Comment (bmancini55) [inline src/exchanges/upbit-client.js]: These numerics should be a string.

File: src/exchanges/upbit-client.spec.js [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+test(
+  "should subscribe and emit ticker events",
+  done => {
+    client.subscribeTicker(market1);
+    client.on("ticker", function tickerHandler(ticker) {
+
+      expect(ticker.fullId).toMatch("Upbit:"+market1.base+"/"+market1.quote);
+      expect(ticker.timestamp).toBeGreaterThan(1531677480465);
+      expect(typeof ticker.last).not.toBe("undefined");
```
Comment (bmancini55) [inline src/exchanges/upbit-client.spec.js]: These should be asserting type "string"

File: src/exchanges/upbit-client.spec.js [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+test(
+  "should subscribe and emit trade events",
+  done => {
+    client.subscribeTrades(market1);
+    client.on("trade", function tradeHandler(trade) {
+
+      try{
+        expect(trade.fullId).toMatch("Upbit:"+market1.base+"/"+market1.quote);
+        expect(trade.exchange).toMatch("Upbit");
```
Comment (bmancini55) [inline src/exchanges/upbit-client.spec.js]: Should have type assertions for strings

File: src/exchanges/upbit-client.js [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+     { ask_price: 1.13,
+       bid_price: 0.94,
+       ask_size: 198013382.3803366,
+       bid_size: 267304509.61145836 } ],
+  stream_type: 'REALTIME' }
+    */
+
+    let remote_id = msg.code;
+    let market = this._level2SnapshotSubs.get(remote_id) || this._level2UpdateSubs.get(remote_id);
+    let asks = msg.orderbook_units.map(p => new Level2Point(p.ask_price, p.ask_size));
```
Comment (bmancini55) [inline src/exchanges/upbit-client.js]: These values should be strings

File: src/exchanges/upbit-client.js [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+     { ask_price: 1.13,
+       bid_price: 0.94,
+       ask_size: 198013382.3803366,
+       bid_size: 267304509.61145836 } ],
+  stream_type: 'SNAPSHOT' }
+    */
+
+    let remote_id = msg.code;
+    let market = this._level2SnapshotSubs.get(remote_id) || this._level2UpdateSubs.get(remote_id);
+    let asks = msg.orderbook_units.map(p => new Level2Point(p.ask_price, p.ask_size));
```
Comment (bmancini55) [inline src/exchanges/upbit-client.js]: These values should be strings

File: src/exchanges/upbit-client.js [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+    let bids = msg.orderbook_units.map(p => new Level2Point(p.bid_price, p.bid_size));
```
Comment (bmancini55) [inline src/exchanges/upbit-client.js]: These values should be strings

File: src/exchanges/upbit-client.js [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+       bid_price: 0.94,
+       ask_size: 198013382.3803366,
+       bid_size: 267304509.61145836 } ],
+  stream_type: 'REALTIME' }
+    */
+
+    let remote_id = msg.code;
+    let market = this._level2SnapshotSubs.get(remote_id) || this._level2UpdateSubs.get(remote_id);
+    let asks = msg.orderbook_units.map(p => new Level2Point(p.ask_price, p.ask_size));
+    let bids = msg.orderbook_units.map(p => new Level2Point(p.bid_price, p.bid_size));
```
Comment (bmancini55) [inline src/exchanges/upbit-client.js]: These values should be strings

File: src/exchanges/upbit-client.js [hunk 8]
```diff
@@ -0,0 +1,320 @@
+const semaphore = require("semaphore");
+const BasicClient = require("../basic-client");
+const BasicMultiClient = require("../basic-multiclient");
+const Ticker = require("../ticker");
+const Trade = require("../trade");
+const Level2Point = require("../level2-point");
+const Level2Snapshot = require("../level2-snapshot");
+const Level2Update = require("../level2-update");
+
+class UpbitClient extends BasicMultiClient {
```
Comment (bmancini55) [inline src/exchanges/upbit-client.js]: Do we need to use BasicMultiClient?

File: src/exchanges/upbit-client.js [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+
+    let { code, trade_timestamp, trade_price, trade_volume, ask_bid, sequential_id } = datum;
+    let market = this._tradeSubs.get(code);
+    let side = (ask_bid === 'BID') ? "buy" : "sell";
+
+    return new Trade({
+      exchange: "Upbit",
+      base: market.base,
+      quote: market.quote,
+      tradeId: sequential_id,
```
[the remark this answers is not in the document]
Comment (panevgen) [reply src/exchanges/upbit-client.js]: We've changed the parsing for this: 
msgs = JSON.parse(msgs.replace(/:(\d+\.{0,1}\d+)(,|\})/g, ':"$1"$2'));

File: src/exchanges/upbit-client.js [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+      side: side,
+      unix: trade_timestamp,
+      price: trade_price,
```
[the remark this answers is not in the document]
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.js [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+      amount: trade_volume,
```
[the remark this answers is not in the document]
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.js (hunk 1 above)
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.spec.js (hunk 2 above)
Comment (panevgen) [reply src/exchanges/upbit-client.spec.js]: FIXED

File: src/exchanges/upbit-client.js (hunk 4 above)
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.js (hunk 5 above)
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.js (hunk 7 above)
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.js (hunk 6 above)
Comment (panevgen) [reply src/exchanges/upbit-client.js]: FIXED

File: src/exchanges/upbit-client.js (hunk 8 above)
Comment (panevgen) [reply src/exchanges/upbit-client.js]: yes, it supports several pairs for a single socket, but in some other way than the others - it overwrites previous values. Here is an example:

Call this:

```
this._wss.send(
        JSON.stringify([{"ticket":"concluded"},{"type":"trade","codes":["BTC_USD", "LTC_USD", "BTT_USD"]}])
      );
```

Returns 3 pairs;

Then adding this:
```
this._wss.send(
        JSON.stringify([{"ticket":"concluded"},{"type":"trade","codes":["XRP_USD"]}])
      );
```
Returns only XRP, not 4 of them.

And for deletion, it's also require to set what "needs to be left", not what "needs to be deleted".

So we should create a new client for this. Please confirm.

Comment (bmancini55) [reply src/exchanges/upbit-client.js]: So it requires sending the active list of markets for both subscribe and unsubscribe events?  
Can't you just iterate on the active markets in the the subscribed market map?  

If you refer to https://github.com/altangent/ccxws/blob/master/src/basic-client.js#L169, the market get added to a map prior to calling the sub-classes subscribe method.  You should be able to construct the complete set of subscribed markets from this.

Comment (panevgen) [reply src/exchanges/upbit-client.js]: Yes, DONE, client is now basic single

**Pull request merged**